### PR TITLE
Added filter docs

### PR DESCRIPTION
**Pull request by samdark**
Added some docs. Please check if these are enough to get started with filters. While putting it together some questions surfaced:

1. Data filters overall aren't bound to REST so probably it's a good idea to move them under `data` namespace. @klimov-paul have you considered it?
2. Not sure it's a good idea to enable filters in REST `IndexAction` by default. My concern is that it could be used for quite complicated conditions that could be exploited to hurt API performance.

| Q             | A
| ------------- | ---
| Is bugfix?    | no
| New feature?  | yes
| Breaks BC?    | no
| Tests pass?   | yes
| Fixed issues  | #12641, #14934

### CODE REVIEW

**Comment by samdark** [comment]
@yiisoft/reviewers would be good if you'll take a look. Thanks.

**Review by dynasource — APPROVED** [review]

File: docs/guide/rest-resources.md [hunk 1]
```diff
@@ -244,4 +244,41 @@ will also include the pagination information by the following HTTP headers:
...
+
+```php
+$filter = new ActiveDataFilter([
+    'searchModel' => 'app\models\PostSearch'
+]);
+
+$filterCondition = null;
+
+// if you prefer JSON in request body, use Yii::$app->request->getBodyParams() below:
+if ($filter->load(\Yii::$app->request->get())) {
```
Comment (bscheshirwork) [inline docs/guide/rest-resources.md]: With `yii\base\Model` and `yii\web\Request` we can use `post(null, null)` and `load(... , '')`
```php
$filter->load(\Yii::$app->request->post(), '')
```
for `json` OR form `post` data...

File: docs/guide/rest-quick-start.md [hunk 2]
```diff
@@ -185,7 +186,11 @@ For example, the URL `http://localhost/users?fields=id,email` will only return t
 
 > Info: You may have noticed that the result of `http://localhost/users` includes some sensitive fields,
 > such as `password_hash`, `auth_key`. You certainly do not want these to appear in your API result.
-> You can and should filter out these fields as described in the [Resources](rest-resources.md) section.
+> You can and should remove these fields from result as described in the [Resources](rest-resources.md) section.
+
+Addionally, you can filter collections like `http://localhost/users?filter[id]=10` or
+`http://localhost/users?filter[email][like]=gmail.com` and sort collections like
+`http://localhost/users?sort=email` or `http://localhost/users?sort=-email`.
```
Comment (cebe) [inline docs/guide/rest-quick-start.md]: should have a link to further reading here, link to `rest-resources.md#filtering-collections`

File: docs/guide/rest-resources.md (hunk 1 above)
Comment (samdark) [reply docs/guide/rest-resources.md]: Of course. Data source could be anything.

Comment (samdark) [reply docs/guide/rest-resources.md]: Do you think it should be mentioned specifically?

File: docs/guide/rest-resources.md [hunk 3]
```diff
@@ -244,4 +244,41 @@ will also include the pagination information by the following HTTP headers:
...
+
+### Filtering collections <span id="filtering-collections"></span>
+
+Since version 2.0.13 Yii provides a facility to filter collections. In case you are using `ActiveController` you can enjoy it
+right away. An example can be found in the [Quick Start](rest-quick-start.md#trying-it-out) section. In case you're
+implementing an endpoint yourself, filtering could be done like the following:
+
+```php
+$filter = new ActiveDataFilter([
+    'searchModel' => 'app\models\PostSearch'
```
Comment (cebe) [inline docs/guide/rest-resources.md]: how does this model look like, there is no standard for this as far as I see.

**Review by cebe — CHANGES_REQUESTED** [review]

**Comment by cebe** [comment]
> Not sure it's a good idea to enable filters in REST IndexAction by default.

does this enable filtering in existing APIs? should not be done imo.

File: docs/guide/rest-resources.md (hunk 1 above)
Comment (bscheshirwork) [reply docs/guide/rest-resources.md]: > // if you prefer JSON in request body, use Yii::$app->request->getBodyParams() below:
This phrase is so concrete. I think we need weaker mention. 
Or more examples...

File: docs/guide/rest-resources.md (hunk 3 above)
Comment (samdark) [reply docs/guide/rest-resources.md]: It's a regular form model. Just public properties and rules. I'll add an example.

**Comment by samdark** [comment]
> does this enable filtering in existing APIs? should not be done imo.

Yes, it does if the API is built around `ActiveController`.

**Comment by samdark** [comment]
@cebe, @dynasource what's your opinion on points 1 and 2 from the PR description?

File: docs/guide/rest-resources.md (hunk 3 above)
Comment (bscheshirwork) [reply docs/guide/rest-resources.md]: in base PR I can see DynamicModel and anonymous function... 
```php
    public function actions()
    {
        return [
            'index' => [
                'class' => 'yii\rest\IndexAction',
                'modelClass' => $this->modelClass,
                'dataFilter' => [
                    'class' => 'yii\rest\ActiveDataFilter',
                    'searchModel' => function () {
                        return (new \yii\base\DynamicModel(['id' => null, 'name' => null]))
                            ->addRule('id', 'integer')
                            ->addRule('name', 'trim')
                            ->addRule('name', 'string');
                    },
                ],
            ],
...
        ];
    }

```
Can process anonymous class `new class()` instead?

**Comment by dynasource** [comment]
@samdark, I tend to agree:
1) it seems generic enough to be put in the data namespace
2) it's already off by default?

File: docs/guide/rest-resources.md (hunk 1 above)
Comment (bscheshirwork) [reply docs/guide/rest-resources.md]: Hm, Interesting realization of `yii\rest\DataFilter::load()`
```php
    /**
     * @inheritdoc
     */
    public function attributes()
    {
        return [
            $this->filterAttributeName
        ];
    }

    /**
     * @inheritdoc
     */
    public function formName()
    {
        return '';
    }
```
@samdark My be we need add link to this in comment?

@see yii\rest\DataFilter::load()
@see yii\rest\DataFilter::attributes()
@see yii\rest\DataFilter::filterAttributeName

**Comment by cebe** [comment]
> Data filters overall aren't bound to REST so probably it's a good idea to move them under data namespace. @klimov-paul have you considered it?

from a quick look into the implementation it seems to me putting it into `yii\data` makes sense.

> Not sure it's a good idea to enable filters in REST IndexAction by default. My concern is that it could be used for quite complicated conditions that could be exploited to hurt API performance.

its not enabled by default as far as I see:

> https://github.com/yiisoft/yii2/blob/ce65e24d1988f00e57ceaa7493632ee4f3aca776/framework/rest/IndexAction.php#L49
> You must setup this field explicitly in order to enable filter processing.

File: docs/guide/rest-resources.md [hunk 4]
```diff
@@ -244,4 +244,64 @@ will also include the pagination information by the following HTTP headers:
...
+]);
+
+$filterCondition = null;
+
+// You may load filters from any source. For example,
+// if you prefer JSON in request body,
+// use Yii::$app->request->getBodyParams() below:
+if ($filter->load(\Yii::$app->request->get())) {
```
Comment (bscheshirwork) [inline docs/guide/rest-resources.md]: Hm, Interesting realization of `yii\rest\DataFilter::load()`
```php
    /**
     * @inheritdoc
     */
    public function attributes()
    {
        return [
            $this->filterAttributeName
        ];
    }

    /**
     * @inheritdoc
     */
    public function formName()
    {
        return '';
    }
```
@samdark My be we need add link to this in comment?
```php
/**
 * You may load filters from any source. For example
 * if you prefer JSON in request body,
 * use Yii::$app->request->getBodyParams() below:
 * @see \yii\rest\DataFilter::load()
 * @see \yii\rest\DataFilter::attributes()
 * @see \yii\rest\DataFilter::filterAttributeName
 */
```

**Review by cebe — APPROVED** [review]

File: docs/guide/rest-resources.md [hunk 5]
```diff
@@ -244,4 +244,64 @@ will also include the pagination information by the following HTTP headers:
...
 An example may be found in the [Quick Start](rest-quick-start.md#trying-it-out) section.
+
+### Filtering collections <span id="filtering-collections"></span>
+
+Since version 2.0.13 Yii provides a facility to filter collections. In case you are using `ActiveController` you can enjoy it
+right away. An example can be found in the [Quick Start](rest-quick-start.md#trying-it-out) section. In case you're
+implementing an endpoint yourself, filtering could be done like the following:
+
+```php
+$filter = new ActiveDataFilter([
```
Comment (bscheshirwork) [inline docs/guide/rest-resources.md]: We need tell more about `DataFilter::attributeMap`. 

This property `attributeMap` (not `searchModel`) can be use (additional) to resolve alias, who we use in resources definition
```php
                    'attributeMap' => [
                        'aliasId' => 'realFieldId',
                    ],
```
```php
class \api\Post extends \common\Post {
    public function fields(){
        $fields = parent::fields();
        unset($fields['realFieldId']);
        $fields['aliasId'] = function () {
            return $this->realFieldId;
        };
        return $fields;
    }
}
```

Comment (bscheshirwork) [reply docs/guide/rest-resources.md]: So... I think we need move $attributeMap to from controller (definition of DataFilter) to the model `searchModel`
```php
class \api\models\PostSearch extends app\models\Post {
    public static function attributeMap(){
         return ['aliasId' => 'realFieldId'];
    }
    public function fields(){
        $fields = parent::fields();
        unset($fields['realFieldId']);
        $fields['aliasId'] = function () {
            return $this->realFieldId;
        };
        return $fields;
    }
}
```
```php
$filter = new ActiveDataFilter([ 
'searchModel' => \api\models\PostSearch::class,
'attributeMap' => [ \api\models\PostSearch::class, 'attributeMap'],
]);
```

i.e. we need closure support of `attributeMap` property too

**Comment by samdark** [comment]
Data filters moved to `data` namespace. Docs adjusted.

File: framework/data/DataFilter.php [hunk 6]
```diff
@@ -388,12 +388,12 @@ public function setErrorMessages($errorMessages)
...
-            'invalidAttributeValueFormat' => Yii::t('yii-rest', "Condition for '{attribute}' should be either a value or valid operator specification."),
-            'operatorRequireAttribute' => Yii::t('yii-rest', "Operator '{operator}' must be used with a search attribute."),
-            'unsupportedOperatorType' => Yii::t('yii-rest', "'{attribute}' does not support operator '{operator}'."),
+            'invalidFilter' => Yii::t('yii-data', 'The format of {filter} is invalid.'),
+            'operatorRequireMultipleOperands' => Yii::t('yii-data', "Operator '{operator}' requires multiple operands."),
+            'unknownAttribute' => Yii::t('yii-data', "Unknown filter attribute '{attribute}'"),
+            'invalidAttributeValueFormat' => Yii::t('yii-data', "Condition for '{attribute}' should be either a value or valid operator specification."),
+            'operatorRequireAttribute' => Yii::t('yii-data', "Operator '{operator}' must be used with a search attribute."),
+            'unsupportedOperatorType' => Yii::t('yii-data', "'{attribute}' does not support operator '{operator}'."),
```
Comment (cebe) [inline framework/data/DataFilter.php]: there are translation files to be renamed:

```
$ find messages/ |grep rest
messages/de/yii-rest.php
messages/hy/yii-rest.php
```

Comment (cebe) [reply framework/data/DataFilter.php]: Also I don't see why we have a separate category for these. why not use `yii` here too? what is different?

Comment (samdark) [reply framework/data/DataFilter.php]: I don't think anything is...

**Review by cebe — APPROVED** [review]

**Pull request merged**